torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp

```diff
@@ -19,17 +19,95 @@ namespace {
 const std::set<c10::Symbol> inplace_ops =
     {aten::append, aten::index_put_, aten::pop, aten::insert, aten::Delete};
 
-bool IsInplaceNode(const Node* n) {
-  if (inplace_ops.find(n->kind()) != inplace_ops.end()) {
-    return true;
-  }
-
-  if (n->kind() == Symbol::fromQualString("onnx::Placeholder") &&
-      n->s(attr::name) == "index_put_") {
-    return true;
+// InplaceConverter defines a set of functions that together enables the conversion
+// from prim::GetAttr, prim::SetAttr, and ATen in-place operators to ONNX out-place operators.
+struct InplaceConverter {
+  InplaceConverter(
+      std::shared_ptr<Graph> graph,
+      MutationRemover* mr,
+      Module* model = nullptr)
+      : graph_(std::move(graph)), mr_(mr), module_(model) {}
+
+  void convertMutationForONNX();
+
+ private:
+  void gatherAttrNameInitialValueMap(
+      Block* block,
+      std::unordered_map<std::string, Value*>& attr_name_value_map,
+      std::unordered_map<Node*, std::string>& attr_node_fullname_map);
+  void replaceAttrWithInplaceOps(
+      Block* block,
+      const std::unordered_map<std::string, Value*>& attr_name_value_map,
+      const std::unordered_map<Node*, std::string>& attr_node_fullname_map);
+
+  void convertInplaceOpsAndTrackAlias();
+  void convertInplaceOpsAndTrackAlias(Block* block);
+
+  void correctAliasReferences();
+  void correctAliasReferences(Block* block);
+  void correctAliasReferences(Node* n);
+
+  void convertGetSetAttrToInplaceOps(Block* block);
+
+  // ValueTracker provides apis to record aliases for a single value,
+  // and to retrieve the correct alias of any given value based on the location in the graph it is used.
+  struct ValueTracker {
+    ValueTracker() : graph_(nullptr) {}
+
+    void init(const std::shared_ptr<Graph>& graph);
+    void recordSetValue(Value* old_v, Value* new_v);
+    Value* findAliasForValueAtNode(Value* v, const Node* n) const;
+
+    std::string toString() const;
+
+   private:
+
+    std::shared_ptr<Graph> graph_;
+
+    // Map from aliases to root value.
```

Comment:
nit: What exactly is an alias? Maybe you could add an explanation here?

Reply:
it is explained in the line below, `A single value can have multiple aliases throughout the graph, created by inplace operators, ...`